plugins/@grouparoo/pardot/src/lib/export/exportProfiles.ts

```diff
@@ -0,0 +1,273 @@
+    await client.deleteProspectById(parseInt(user.destinationId));
+  }
+};
+
+// update these users by destinationId
+const updateByDestinationIds: BatchMethodUpdateByDestinationIds = async ({
+  client,
+  users,
+  config,
+}) => {
+  const { cacheData } = config.data;
+  await updateProspectsHelper(client, users, cacheData);
+};
+
+// usually this is creating them. ideally upsert. set the destinationId on each when done
+const createByForeignKeyAndSetDestinationIds: BatchMethodCreateByForeignKeyAndSetDestinationIds = async ({
+  client,
+  users,
+  config,
+}) => {
+  const { cacheData } = config.data;
+
+  await updateProspectsHelper(client, users, cacheData);
+
+  for (let user of users) {
+    // We're not using destinationIds for created prospects, but grouparoo requires it to be set
+    // and the batch upsert does not return IDs, so we'll just set it to the foreignKey
+    if (!user.destinationId) {
+      user.destinationId = user.foreignKeyValue;
+    }
+  }
+};
+
+async function updateProspectsHelper(
+  client: PardotClient,
+  users: BatchExport[],
+  cacheData: PardotCacheData
+) {
+  if (users.length === 0) return; // Nothing to do
+
+  // Sometimes we may get "414 - URI Too Long" from Pardot.
+  // In extremely long URIs, 414 won't be returned and "400 - Bad Request" will be returned instead.
+  // In these cases, we recursively split into two batches...
+  try {
+    await updateProspects(client, users, cacheData);
+  } catch (error) {
+    if (error.response?.status === 414 || error.response?.status === 400) {
+      if (users.length === 1) {
+        users[0].error = new Error(
+          `${error.response.status} - ${error.response.statusText}`
+        );
+        return;
+      }
+
+      const half = Math.ceil(users.length / 2);
+      await updateProspectsHelper(client, users.slice(0, half), cacheData);
+      await updateProspectsHelper(client, users.slice(half), cacheData);
+    } else {
+      throw error;
+    }
+  }
+}
+
+async function updateProspects(
+  client: PardotClient,
+  users: BatchExport[],
+  cacheData: PardotCacheData
+) {
+  const input: any[] = [];
+  for (const exportedProfile of users) {
+    const payload = await buildPayload(client, exportedProfile, cacheData);
+    input.push(payload);
+  }
+
+  const errors = await client.batchUpsertProspects(input);
+  for (let errorIdx in errors) {
+    const resultIdx = parseInt(errorIdx);
+    const user = users[resultIdx];
+    user.error = new Error(errors[errorIdx]);
+  }
+}
+
+async function buildPayload(
+  client: PardotClient,
+  exportedProfile: BatchExport,
+  cacheData: PardotCacheData
+) {
+  const {
+    destinationId,
+    oldProfileProperties,
+    newProfileProperties,
+    oldGroups,
+    newGroups,
+    foreignKeyValue,
+  } = exportedProfile;
+
+  const payload: any = {};
+
+  payload.email = foreignKeyValue; // email
+  if (destinationId) {
+    payload.id = destinationId;
+  }
+
+  if (!payload.email) {
```

Comment:
I think we actually check for this and `throw` in `findAndSetDestinationIds` (because it using the email)

Reply:
actually maybe the helper does that. let's check. either way out form here.

Reply:
You're right, the helper does that for the foreignKey. Added this here because it's a required field, in case we were going to support different foreignKeys other than email. Removing it for now.